module/VuFind/src/VuFind/Controller/OAuth2ControllerFactory.php

```diff
@@ -0,0 +1,289 @@
+ * @link     https://vufind.org Main Site
+ */
+namespace VuFind\Controller;
+
+use Interop\Container\ContainerInterface;
+use Interop\Container\Exception\ContainerException;
+use Laminas\ServiceManager\Exception\ServiceNotCreatedException;
+use Laminas\ServiceManager\Exception\ServiceNotFoundException;
+use League\OAuth2\Server\AuthorizationServer;
+use League\OAuth2\Server\Grant\AuthCodeGrant;
+use League\OAuth2\Server\Grant\RefreshTokenGrant;
+use League\OAuth2\Server\ResourceServer;
+use League\OAuth2\Server\ResponseTypes\BearerTokenResponse;
+use League\OAuth2\Server\ResponseTypes\ResponseTypeInterface;
+use OpenIDConnectServer\ClaimExtractor;
+use OpenIDConnectServer\Entities\ClaimSetEntity;
+use OpenIDConnectServer\IdTokenResponse;
+use VuFind\Config\Locator;
+use VuFind\Db\Table\AccessToken;
+use VuFind\OAuth2\Repository\AccessTokenRepository;
+use VuFind\OAuth2\Repository\AuthCodeRepository;
+use VuFind\OAuth2\Repository\ClientRepository;
+use VuFind\OAuth2\Repository\IdentityRepository;
+use VuFind\OAuth2\Repository\RefreshTokenRepository;
+use VuFind\OAuth2\Repository\ScopeRepository;
+
+/**
+ * OAuth2 controller factory.
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class OAuth2ControllerFactory extends AbstractBaseFactory
+{
+    /**
+     * Service manager
+     *
+     * @var ContainerInterface
+     */
+    protected $container;
+
+    /**
+     * OAuth2 configuration
+     *
+     * @var array
+     */
+    protected $oauth2Config;
+
+    /**
+     * Access token table
+     *
+     * @var AccessToken
+     */
+    protected $accessTokenTable;
+
+    /**
+     * Protected scopes in OpenID Connect
+     *
+     * @var array
+     */
+    protected $openIdConnectProtectedScopes = [
+        'profile',
+        'email',
+        'address',
+        'phone'
+    ];
+
+    /**
+     * Create an object
+     *
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $oauth2Config = $yamlReader->get('OAuth2Server.yaml');
```

Comment:
You could save a line of code by assigning directly to `$this->oauth2Config` here and then changing the access to `$oauth2Config` below.

Reply:
Fixed.